pkg/util/context/plancache.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package context
+
+import (
+	"sync"
+
+	"github.com/pingcap/errors"
+)
+
+// PlanCacheType is the flag of plan cache
+type PlanCacheType int
+
+const (
+	// DefaultNoCache no cache
+	DefaultNoCache PlanCacheType = iota
+	// SessionPrepared session prepared plan cache
+	SessionPrepared
+	// SessionNonPrepared session non-prepared plan cache
+	SessionNonPrepared
+)
+
+// PlanCacheTracker PlanCacheTrackerf `PlanCacheTracker`.
+type PlanCacheTracker struct {
+	UseCache            bool
+	CacheType           PlanCacheType
+	ForcePlanCache      bool // force the optimizer to use plan cache even if there is risky optimization, see #49736.
+	alwaysWarnSkipCache bool
+
+	warnHandler WarnHandler
+}
+
+// ForceSetSkipPlanCache forces to skip plan cache.
+func (h *PlanCacheTracker) ForceSetSkipPlanCache(err error) {
+	if h.CacheType == DefaultNoCache {
+		return
+	}
+	h.setSkipPlanCache(err)
+}
+
+// SetSkipPlanCache sets to skip plan cache.
+func (h *PlanCacheTracker) SetSkipPlanCache(err error) {
+	if !h.UseCache {
+		return
+	}
+
+	if h.ForcePlanCache {
+		h.warnHandler.AppendWarning(errors.NewNoStackErrorf("force plan-cache: may use risky cached plan: %s", err.Error()))
+		return
+	}
+
+	h.setSkipPlanCache(err)
+}
+
+func (h *PlanCacheTracker) setSkipPlanCache(err error) {
+	h.UseCache = false
+	switch h.CacheType {
+	case DefaultNoCache:
+		h.warnHandler.AppendWarning(errors.NewNoStackError("unknown cache type"))
+	case SessionPrepared:
+		h.warnHandler.AppendWarning(errors.NewNoStackErrorf("skip prepared plan-cache: %s", err.Error()))
+	case SessionNonPrepared:
+		if h.alwaysWarnSkipCache {
```

Comment:
Is it better to rename it to `alwaysWarnSkipNoPreparedCache` because it only works on `SessionNonPrepared` ?

Reply:
I think the semantic is to warn in any cases, but for other cases there is already a warn so it only applies to `SessionNonPrepared`.

It's used in `explain format='plan_cache'  xxx`. It should give a warning to show the status of any kinds of plan cache.